# EXERCISE 4.5 DATA CONSISTENCY CHECKS

## Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

## Importing data

In [2]:
# Assign path
path = r'C:\Users\fiyin\OneDrive\Documents\InstaCart Basket Analysis'

In [3]:
# Import products.csv
df_prods = pd.read_csv(os.path.join(path, '02. Data', 'Original data', 'products.csv'), index_col = False)

In [4]:
# Import orders_wrangled.csv
df_ords = pd.read_csv(os.path.join(path, '02. Data', 'Prepared data', 'orders_wrangled.csv'), index_col = False)

## Handling mixed-type data

In [6]:
# Create a test dataframe
df_test = pd.DataFrame()

In [7]:
# Create a mixed type column in the test dataframe
df_test['mix'] = ['a', 'b', 1, True]

In [8]:
# Viewing test dataframe
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# Checking for mixed data types in the test dataframe
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [16]:
# Changing datatype of 'mix' column in test dataframe
df_test['mix'] = df_test['mix'].astype('str')

## Handling missing data

In [19]:
# Checking for missing data in products.csv
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [20]:
# Create a subset dataframe for missing values in 'product_name' column
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [21]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [30]:
df_prods.shape

(49693, 5)

In [29]:
# Creating new dataframe 'df_prods_clean' excluding missing values from 'product_name' column
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [31]:
df_prods_clean.shape

(49677, 5)

## Handling duplicates

In [33]:
# Create subset dataframe 'df_dups' containing only duplicate rows in df_prods_clean
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [34]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [35]:
# Deleting duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [36]:
df_prods_clean_no_dups.shape

(49672, 5)

## Exporting data

In [37]:
# Exporting checked products.csv data as products_checked.csv
df_prods_clean_no_dups.to_csv(os.path.join(path, '02. Data', 'Prepared data', 'products_checked.csv'))

# TASK 4.5 DATA CONSISTENCY CHECKS

### 2. Run the df.describe() function on your df_prods dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [41]:
# Summary statistics for products.csv
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [96]:
df_prods[df_prods['prices'] == 99999]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

In [105]:
df_prods[df_prods['product_name'] == '2% Reduced Fat Milk']

product_id         product_name  aisle_id  department_id  prices
23910       23909  2% Reduced Fat Milk        84             16     9.2

#### Answer to Q2: The maximum value for the prices column looks off. It is most likely an error. I found a similar item with the same aisle_id and department_id, with a more realistic price: 9.2, though the product_id is different.

### 3. Check for mixed-type data in your df_ords dataframe.

In [43]:
# Checking for mixed-type data in df_ords columns
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

### 4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

#### From Q3 above, there are no mixed-type data in df_ords dataframe

### 5. Run a check for missing values in your df_ords dataframe.
In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [8]:
# Checking for missing values in df_ords
df_ords.isnull().sum()

Unnamed: 0                      0
order_id                        0
user_id                         0
number_of_items_ordered         0
order_day_of_week               0
order_hour_of_day               0
days_since_prior_order     206209
dtype: int64

In [92]:
# View of subset with only missing values from the 'days_since_prior_order' column
df_ords[df_ords['days_since_prior_order'].isnull() == True]

order_id  user_id  number_of_items_ordered  order_day_of_week  \
0         2539329        1                        1                  2   
11        2168274        2                        1                  2   
26        1374495        3                        1                  1   
39        3343014        4                        1                  6   
45        2717275        5                        1                  3   
...           ...      ...                      ...                ...   
3420930    969311   206205                        1                  4   
3420934   3189322   206206                        1                  3   
3421002   2166133   206207                        1                  6   
3421019   2227043   206208                        1                  1   
3421069   3154581   206209                        1                  3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

In [90]:
# Checking statistics of user_id column in the df_ords dataframe. 
df_ords['user_id'].describe()

count    3.421083e+06
mean     1.029782e+05
std      5.953372e+04
min      1.000000e+00
25%      5.139400e+04
50%      1.026890e+05
75%      1.543850e+05
max      2.062090e+05
Name: user_id, dtype: float64

#### Answer to Q5: There are 206209 missing values in the 'days_since_prior_order' column. If there are a total of 206209 customers, then it is likely that the missing values refer to the first ever order for each customer.

### 6. Address the missing values using an appropriate method.
In a markdown cell, explain why you used your method of choice.

####  Answer to Q6: I decided to leave the missing values as they are. Adding a string flag variable would mean a mixed-type data in the colum. A numeric flag variable may skew future analysis.

In [144]:
df_ords.describe()

order_id       user_id  number_of_items_ordered  order_day_of_week  \
count  3.421083e+06  3.421083e+06             3.421083e+06       3.421083e+06   
mean   1.710542e+06  1.029782e+05             1.715486e+01       2.776219e+00   
std    9.875817e+05  5.953372e+04             1.773316e+01       2.046829e+00   
min    1.000000e+00  1.000000e+00             1.000000e+00       0.000000e+00   
25%    8.552715e+05  5.139400e+04             5.000000e+00       1.000000e+00   
50%    1.710542e+06  1.026890e+05             1.100000e+01       3.000000e+00   
75%    2.565812e+06  1.543850e+05             2.300000e+01       5.000000e+00   
max    3.421083e+06  2.062090e+05             1.000000e+02       6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

### 7. Run a check for duplicate values in your df_ords data.

In [9]:
# Checking for duplicates in the df_ords data
df_ords.duplicated().sum()

0

#### There are no duplicates in the df_ords data

### 9. Export your final, cleaned df_ords data as a “.csv” file in your “Prepared Data” folder and give them appropriate, succinct names.

In [10]:
# Export checked orders data
df_ords.to_csv(os.path.join(path, '02. Data', 'Prepared data', 'orders_checked.csv'))